netlify/next-runtime
# fix: correctly handle ISR for appDir pages

### Summary
Data routes are handled differently for appDir pages. Instead of a JSON file in `/_next/data`, the request is made to the same path, passing the `RSC` header. This returns the RSC data to stream the update. The fact this varies on a header means we can't use simple ODB handling for these pages, because it will lead to errors like #1844, where the RSC data is returned when HTML is requested, or vice versa. 

This PR fixes this by adding an edge function which routes appDir requests that include the `rsc` header to the appropriate `.rsc` path instead. The original request path is put into a header, which then replaces the `.rsc` path in the handler. These are real files but not normally requested directly. The reason for doing this is so that these requests have unique URLs, meaning they're cached separately as ODBs. 

There's a final extra part that is needed, which is to handle trailing slashes. If we normalise all of the requests to the matching `.rsc` file, then there will be no difference to requests with and without the trailing slash. This is a problem, because depending on whether the version with or without the slash is requested first, we may cache the redirect instead of the actual data (or vice versa). To handle this, we do a bit of trickery. If the request has a trailing slash, we include a trailing slash on the rewritten `.rsc`path. For example, while an rsc request to `/blog` will be rewritten to `/blog.rsc`, a request for `/blog/` (with trailing slash) will be rewritten to `/blog.rsc/`. This means that it will be treated as a separate ODB path, and will be cached differently.

Basically, all of this is to implement `Vary` in ODBs. 🤷🏻 

### Test plan

1. Visit [an appDir page on the default demo site](https://deploy-preview-1855--netlify-plugin-nextjs-demo.netlify.app/blog/erica/). Navigate around, checking in the network panel that SPA data requests are correctly returning `application/octet-stream` RSC data. 
2. Load any of the `/blog` pages directly and ensure that HTML is returned
3. Try loading them with and without the trailing slash and see if they redirect correctly.

### Relevant links (GitHub issues, Notion docs, etc.) or a picture of cute animal
Fixes #1844

![capy wants an apple](https://user-images.githubusercontent.com/213306/212292611-cd822b62-48c8-414f-9f7d-210123e67095.png)

### Standard checks:




---

🧪 Once merged, make sure to update the version if needed and that it was published correctly.
